### PR TITLE
Testnet: add new container for build

### PR DESCRIPTION
**Pull request by MoroccanMalinois**
---
**By submitting this pull-request, I confirm the following:**

- I have read and understood the contributor guide in [kovri-docs](https://github.com/monero-project/kovri-docs).
- I have checked that another pull-request for this purpose does not exist.
- I have considered and confirmed that this submission will be valuable to others.
- I accept that this submission may not be used and that this pull-request may be closed by the will of the maintainer.
- I give this submission freely under the BSD 3-clause license.
---

To avoid rebuilding the docker image for each git revision, the user can now choose to "Use binaries from repo" during testnet creation.

To rebuild the binaries:

    testnet.sh exec make

Or get a shell first:

    testnet.sh exec sh

and then

     make

### CODE REVIEW

File: contrib/docker/testnet.sh [hunk 1]
```diff
@@ -74,6 +72,28 @@ set_repo()
...
+      [yY])
+        KOVRI_USE_REPO_BINS=true
+        ;;
+      *)
+        KOVRI_USE_REPO_BINS=false
+      ;;
+     esac
+  fi
+
+  mount_repo_bins=""
```
Comment (anonimal) [inline contrib/docker/testnet.sh]: >mount_repo_bins=""

I don't think this is necessary. Did you find this necessary?

**Comment by MoroccanMalinois** [comment]
In addition, in the new dockerfile, maybe we could also add gdb (i took the liberty to add iptables as my next PR will require it). 

Just an fyi, whatever the size of the image, it doesn't change the size of instances (as it uses [AUFS](https://en.wikipedia.org/wiki/Aufs)), so the impact of the docker image size on testnet is quite limited.

File: contrib/docker/testnet.sh [hunk 2]
```diff
@@ -74,6 +72,28 @@ set_repo()
...
+  if [[ "$KOVRI_USE_REPO_BINS" = true ]]; then
+    mount_repo_bins="-v ${KOVRI_REPO}/build/kovri:/usr/bin/kovri -v ${KOVRI_REPO}/build/kovri-util:/usr/bin/kovri-util"
+
```
Comment (anonimal) [inline contrib/docker/testnet.sh]: If we're going to start bool'ing, how about we create bool vars for building? This [bool'ing] can then be applied elsewhere so we never have to enter keyboard input once they are all set.

What do you think? If we apply elsewhere, I think it should be in a separate commit.

File: contrib/docker/testnet.sh [hunk 3]
```diff
@@ -101,7 +121,16 @@ set_image()
...
-  read_input "Build Kovri Docker image? [$KOVRI_IMAGE]" NULL "docker build -t $KOVRI_IMAGE $KOVRI_REPO"
+  # Select Dockerfile
+  local _default_dockerfile="Dockerfile"
+  if [[ -z $KOVRI_DOCKERFILE ]]; then
+    KOVRI_DOCKERFILE=${_default_dockerfile}
+    read_input "Change Dockerfile?: [KOVRI_DOCKERFILE=${KOVRI_DOCKERFILE}]" KOVRI_DOCKERFILE
+  fi
+
+  local _dockerfile_path="${KOVRI_REPO}/contrib/docker/${KOVRI_DOCKERFILE}"
+
```
Comment (anonimal) [inline contrib/docker/testnet.sh]: Same here.

File: contrib/docker/testnet.sh [hunk 4]
```diff
@@ -314,6 +343,17 @@ Destroy()
...
 
+Exec()
+{
+  docker run -i -t \
+    --rm \
+    -v $KOVRI_REPO:/home/kovri/kovri \
+    -w /home/kovri/kovri \
+    $KOVRI_IMAGE \
+    $@
+  catch "Failed to run a shell"
```
Comment (anonimal) [inline contrib/docker/testnet.sh]: >"Failed to run a shell"

But this could be a failed docker run too?

**Comment by anonimal** [comment]
Hi @MoroccanMalinois, I just saw these:

>In addition, in the new dockerfile, maybe we could also add gdb (i took the liberty to add iptables as my next PR will require it).

Ok, sounds good. Since it's dev-only, let's do what we need.

>Just an fyi, whatever the size of the image, it doesn't change the size of instances (as it uses AUFS EDIT: or OverlayFS now ), so the impact of the docker image size on testnet is quite limited.

I'm more concerned about how we manage logging. Even without debug levels, there's a lot of writing. I have some ideas on potential resolutions, will talk more in `#kovri-dev`.

**Comment by MoroccanMalinois** [comment]
Thanks @anonimal. Ready for review

File: contrib/docker/testnet.sh [hunk 5]
```diff
@@ -74,6 +100,19 @@ set_repo()
...
+set_bins()
+{
+  read_bool_input "Use binaries from repo?" KOVRI_USE_REPO_BINS ""
+
+  if [[ $KOVRI_USE_REPO_BINS = true ]];then
+    echo "Using binaries in ${KOVRI_REPO}/build"
+    mount_repo_bins="-v ${KOVRI_REPO}/build/kovri:/usr/bin/kovri \
+      -v ${KOVRI_REPO}/build/kovri-util:/usr/bin/kovri-util"
+
+    read_bool_input "Build repo binaries?" KOVRI_BUILD_REPO_BINS "Exec make"
```
Comment (anonimal) [inline contrib/docker/testnet.sh]: Re: our debugging in `#kovri-dev`, any objection to making this a `release-static` build?

File: contrib/docker/testnet.sh [hunk 6]
```diff
@@ -74,6 +100,19 @@ set_repo()
   fi
 }
 
+set_bins()
+{
+  read_bool_input "Use binaries from repo?" KOVRI_USE_REPO_BINS ""
```
Comment (anonimal) [inline contrib/docker/testnet.sh]: Can you add a note to the question (or somewhere) about ensuring that the binaries are built statically if not built within the container?

File: contrib/docker/testnet.sh [hunk 7]
```diff
@@ -40,12 +65,13 @@ Prepare()
 
   # Cleanup for new testnet
   if [[ $KOVRI_WORKSPACE || $KOVRI_NETWORK ]]; then
-    read_input "Kovri testnet environment detected. Attempt to destroy previous testnet?" NULL cleanup_testnet
+    read_bool_input "Kovri testnet environment detected. Attempt to destroy previous testnet?" KOVRI_CLEANUP cleanup_testnet
```
[the remark this answers is not in the document]
Comment (MoroccanMalinois) [reply contrib/docker/testnet.sh]: yes. you can set it to `true` or `false`, and the script will destroy or not without asking.
At least, it's what it's supposed to do :)

[the remark this answers is not in the document]
Comment (anonimal) [reply contrib/docker/testnet.sh]: >  if [ "${!2}" = true ];then
>    $3
>  fi

Sorry, old comment
